src/crdt.rs

```diff
@@ -235,6 +240,37 @@ impl Crdt {
                 self.table[&v.id].version
             );
         }
+        //update the liveness table
+        let now = timestamp();
+        *self.alive.entry(v.id).or_insert(now) = now;
+    }
+
+    /// purge old validators
+    /// TODO: we need a robust membership protocol
+    /// http://asc.di.fct.unl.pt/~jleitao/pdf/dsn07-leitao.pdf
+    /// challenging part is that we are on a permissionless network
+    pub fn purge(&mut self, now: u64) {
+        //wait for 4x as long as it would randomly take to reach our node
+        //assuming everyone is waiting the same amount of time as this node
+        let limit = self.table.len() as u64 * GOSSIP_SLEEP_MILLIS * 4;
+        let purge_set: Vec<PublicKey> = self.alive
+            .iter()
+            .filter_map(|(k, v)| {
```

Comment:
You might want to try something like `(ref k, ref v)` or `&(k, v)` to get rid of all those dereferences below. Clippy would probably tell you which.